### PR TITLE
KAFKA-7632: Allow fine-grained configuration for compression

### PR DESCRIPTION
**Pull request by dongjinleekr**
This PR implements [KIP-390: Add producer option to adjust compression level](https://cwiki.apache.org/confluence/display/KAFKA/KIP-390%3A+Add+producer+option+to+adjust+compression+level).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by dongjinleekr** [comment]
Please note that this is a draft implementation. It will be updated following the update of zstd-jni by @luben. (see: luben/zstd-jni#80)

File: clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java [hunk 1]
```diff
@@ -166,6 +166,12 @@
                                                        + " values are <code>none</code>, <code>gzip</code>, <code>snappy</code>, <code>lz4</code>, or <code>zstd</code>. "
                                                        + "Compression is of full batches of data, so the efficacy of batching will also impact the compression ratio (more batching means better compression).";
 
+    /** <code>compression.level</code> */
+    public static final String COMPRESSION_LEVEL_CONFIG = "compression.level";
+    private static final String COMPRESSION_LEVEL_DOC = "The compression level for all data generated by the producer. The default level and valid value is up to "
+                                                        + "compression.type. (<code>none</code>, <code>snappy</code>: not available. <code>gzip</code>: 1~9 (default: 1). "
+                                                        + "<code>lz4</code>: 4~7 (default: 4). <code>zstd</code>: 1~22 (default: 3).";
```
Comment (bobrik) [inline clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java]: What about negative (fast) ones for zstd?

**Comment by luben** [comment]
I just shipped zstd-jni-1.3.7-2 that adds methods to query min/max supported compression levels.

File: clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java (hunk 1 above)
Comment (dongjinleekr) [reply clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java]: I will be updated with zstd-jni-1.3.7-2 accordingly. It was just released about 4 hours from now.

Comment (luben) [reply clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java]: @dongjinleekr , yes, I noticed that you also want it and I pushed a release.

**Comment by dongjinleekr** [comment]
I just applied zstd-jni-1.3.7-2. Thanks again @luben!!

**Comment by dongjinleekr** [comment]
And there is another issue here; as of now, Kafka's checkstyle rule disallows more than 12 parameters for any method. However, with introducing a new configuration entity, `MemoryRecords#builder` method requires 13 parameters - it is why the automated check fails. To complete this PR, this problem should also be solved.

**Comment by dongjinleekr** [comment]
And there is another issue here; as of now, Kafka's checkstyle rule disallows more than 13 parameters for any method. However, with introducing a new configuration entity, the constructor of `MemoryRecordsBuilder` (line 86, 160) requires 14 parameters - it is why the automated check fails. To complete this PR, this problem should also be solved.

**Comment by dongjinleekr** [comment]
retest this please

**Comment by dongjinleekr** [comment]
retest this please

**Comment by dongjinleekr** [comment]
Here is the re-implementation of KIP-390, rebased against the latest trunk. The commits as of present consist like the following:

1. Refactor the existing code.
    1. Replace current `MemoryRecords#builder` implementation with a new one with a fluent method call. It enables `MemoryRecords#builder` to support more than 13 parameters, also making the rebase against the latest trunk easier.
    2. Change `BrokerCompressionCodec#getTargetCompressionCodec` to return `CompressionType` instead of `CompressionCodec`. It enables the introduction of a new class, `CompressionConfig`. (A new test method for validation was also added.)
2. Extend the compression functionality to support compression level and buffer size.
    1. Refactor `KafkaLZ4BlockOutputStream`, `KafkaLZ4Test` and add support for compression level, block size configuration to lz4.
    2. Add `GZipOutputStream` to support compression level, block size for Gzip.
    3. Add `CompressionConfig` which encapsulates compression type, level, and buffer size, and make Producer and Broker use it.
3. Implement two alternative public interfaces that are under discussion:
    - Type A: Adds `compression.level` and `compression.buffer.size`.
    - Type B: Adds `compression.config` (like `listener.security.protocol.map` or `max.connections.per.ip.overrides`)

Since the public interface is not settled yet, this PR currently includes the common commits by step 2. For the final ones, please refer the following branches:

- Type A: https://github.com/dongjinleekr/kafka/tree/feature/KAFKA-7632-a
- Type B: https://github.com/dongjinleekr/kafka/tree/feature/KAFKA-7632-b

**Comment by dongjinleekr** [comment]
Retest this please.

**Comment by dongjinleekr** [comment]
Retest this please.

**Comment by dongjinleekr** [comment]
Closes this PR in preference of a [new one](https://github.com/apache/kafka/pull/10826).

**Pull request closed without merging**